todo/repositories/role_repository.py

```diff
@@ -0,0 +1,123 @@
+
+from todo.models.role import RoleModel
+from todo.repositories.common.mongo_repository import MongoRepository
+from todo.constants.role import RoleType, RoleScope
+
+logger = logging.getLogger(__name__)
+
+
+class RoleRepository(MongoRepository):
+    collection_name = RoleModel.collection_name
+
+    @classmethod
+    def list_all(cls, filters: Optional[Dict[str, Any]] = None) -> List[RoleModel]:
+        roles_collection = cls.get_collection()
+
+        query = {}
+        if filters:
+            if "is_active" in filters:
+                query["is_active"] = filters["is_active"]
+            if "type" in filters:
+                query["type"] = filters["type"]
+            if "scope" in filters:
+                query["scope"] = filters["scope"]
+
+        roles_cursor = roles_collection.find(query)
+        roles = []
+
+        for role_doc in roles_cursor:
+            try:
+                role_model = cls._document_to_model(role_doc)
+                roles.append(role_model)
+            except Exception as e:
+                logger.error(f"Error converting role document to model: {e}")
+                logger.error(f"Document: {role_doc}")
+                continue
+
+        return roles
+
+    @classmethod
+    def _document_to_model(cls, role_doc: dict) -> RoleModel:
+        if "type" in role_doc and isinstance(role_doc["type"], str):
+            role_doc["type"] = RoleType(role_doc["type"])
+
+        if "scope" in role_doc and isinstance(role_doc["scope"], str):
+            role_doc["scope"] = RoleScope(role_doc["scope"])
+
+        return RoleModel(**role_doc)
+
+    @classmethod
+    def create(cls, role: RoleModel) -> RoleModel:
+        roles_collection = cls.get_collection()
+
+        existing_role = roles_collection.find_one({"name": role.name})
+        if existing_role:
+            raise ValueError(f"Role with name '{role.name}' already exists")
+
+        role.created_at = datetime.now(timezone.utc)
+        role.updated_at = None
+
+        role_dict = role.model_dump(mode="json", by_alias=True, exclude_none=True)
+        insert_result = roles_collection.insert_one(role_dict)
+
+        role.id = insert_result.inserted_id
+        return role
+
+    @classmethod
+    def get_by_id(cls, role_id: str) -> Optional[RoleModel]:
+        roles_collection = cls.get_collection()
+        role_data = roles_collection.find_one({"_id": ObjectId(role_id)})
+        if role_data:
+            return cls._document_to_model(role_data)
+        return None
+
+    @classmethod
+    def update(cls, role_id: str, update_data: dict) -> Optional[RoleModel]:
+        try:
+            obj_id = ObjectId(role_id)
+        except InvalidId:
+            return None
+
+        if "name" in update_data:
+            existing_role = cls.get_by_name(update_data["name"])
+            if existing_role and str(existing_role.id) != role_id:
+                raise ValueError(f"Role with name '{update_data['name']}' already exists")
```

Comment:
- can we use the `RoleAlreadyExistsException` from `todo/exceptions/role_exceptions.py`?

Reply:
done